src/dbnode/storage/flush.go

```diff
@@ -159,26 +170,48 @@ func (m *flushManager) Flush(startTime xtime.UnixNano) error {
 		multiErr = multiErr.Add(fmt.Errorf("error rotating commitlog in mediator tick: %v", err))
 	}
 
-	if err = m.indexFlush(namespaces); err != nil {
+	indexFlushes, err := m.indexFlush(namespaces)
+	if err != nil {
 		multiErr = multiErr.Add(err)
 	}
 
-	return multiErr.FinalError()
+	err = multiErr.FinalError()
+
+	// Mark all flush states at the very end to ensure this
+	// happens after both data and index flushing.
+	// TODO: if this matching approach of data + index ns+shard+time is needed, then reimplement this
+	// more efficiently w/ hashing for constant lookup cost.
+	for _, f := range flushes {
+		for _, s := range f.shardFlushes {
+			for _, ff := range indexFlushes {
```

Comment:
FYI originally I was just using `flushes` and not checking it against `indexFlushes` assuming that anything that could have been "data" flushed must have been "index" flushes since it happens after. That assumption maybe isn't correct, though, and so now this is ensuring that marking a flush requires that the ns+shard+time happened in the warmFlush + indexFlush path.

This is just hacked together atm to test it out quickly and so will need to refactor if we do need this.

Reply:
Another part I suspect might be incorrect here is these time comparisons - given that index block size and block size usually differ?

Reply:
Out of date - since have converted to using maps + using either index or data as the signal for marking.